README

Help on actions:
-------------------

* Compile and install
* Build debian package
* Make documentation
* Add entry to changelog


Compile and install
-------------------
    To build libsphinxclient and install it to your system from source
    distribution run following commands from the topmost directory
    of sphinxclient.

        # autoreconf --install
        # ./configure
        # make clean
        # make
        # make install


Build debian package
--------------------

    In the topmost directory run these comamnds:

        # autoreconf --install
        # dpkg-buildpackage -rfakeroot

    The distribution packages are placed in parent directory (cd ..)


Make documentation
------------------

    To make documentation, you must have doxygen installed. Then, you can run

        # ./make-documentation.sh

    in the topmost directory. The result can be found in ./doc directory.


Add entry to changelog
----------------------

    # dch -v <release version>-<revision>

    Then edit the entry in changelog.



debian/changelog
libsphinxclient (2.1.4) unstable; urgency=low

  * Added unix domain socket config option.

 -- Jan Cermak <jan.cermak@firma.seznam.cz>  Mon, 24 Apr 2017 16:08:14 +0200

libsphinxclient (2.1.3-2) unstable; urgency=low

  * pkg-config fix

 -- Ondrej Kocian <ondrej.kocian@firma.seznam.cz>  Mon, 02 Dec 2013 09:11:15 +0100

libsphinxclient (2.1.3-1) unstable; urgency=low

  * missing pkg-config
  * usleep missing header on wheezy

 -- Ondrej Kocian <ondrej.kocian@firma.seznam.cz>  Fri, 29 Nov 2013 19:16:55 +0100

libsphinxclient (2.1.3) unstable; urgency=low

  * Fixed broken MVA64 attributes fetching.
  * Added missing string value type name.
  * Debian packaging - allow applications to depend on a specific
    libsphinxclient package version.

 -- Jan Kirschner <jan.kirschner@firma.seznam.cz>  Thu, 21 Mar 2013 14:34:36 +0100

libsphinxclient (2.1.2) unstable; urgency=low

  * Added a function to escape query string to be safe for extended query
    syntax.

 -- Jan Kirschner <jan.kirschner@firma.seznam.cz>  Thu, 07 Mar 2013 14:18:26 +0100

libsphinxclient (2.1.1) unstable; urgency=low

  * Fix errorneous copyconstructor of SearchConfig_t.

 -- Jan Kirschner <jan.kirschner@firma.seznam.cz>  Wed, 30 Jan 2013 14:41:14 +0100

libsphinxclient (2.1.0) unstable; urgency=low

  * Support string attributes and 64bit multi-attributes.
  * Add newest ranking modes except.
  * Remove support for sphinx search older than 0.9.9.
  * Rewrite SearchConfig_t interface to use dptr.

 -- Jan Kirschner <jan.kirschner@firma.seznam.cz>  Mon, 21 Jan 2013 10:48:15 +0100

libsphinxclient (2.0.1) unstable; urgency=low

  * Interface for fetching filter data from search config.
  * Method for validating objects of Value_t.

 -- Jan Kirschner <jan.kirschner@firma.seznam.cz>  Wed, 21 Nov 2012 12:21:01 +0100

libsphinxclient (2.0.0) unstable; urgency=low

  * Binary incompatible, new version of debian package
  * Obvious unnecessary things removed from public interface (filter data)
  * Added configurable support for connection retry
  * All networking handled newly by new layer QueryMachine_t (polling state
    machine)

 -- Vaclav Urbanek <vaclav.urbanek@firma.seznam.cz>  Thu, 20 Sep 2012 08:49:21 +0200

libsphinxclient (1.4.0) unstable; urgency=low

  * added new type of multiquery, capable of optimizing subqueries,
    grouping them and send groups simultaneously to searchd 

 -- Vaclav Urbanek <vaclav.urbanek@firma.seznam.cz>  Tue, 11 Sep 2012 14:58:43 +0200

libsphinxclient (1.3.3-1) unstable; urgency=low

  * Fixed polling sockets ready for reading.

 -- Daniel Fiser <daniel.fiser@firma.seznam.cz>  Thu, 28 Jun 2012 11:06:06 +0200

libsphinxclient (1.3.2-1) unstable; urgency=low

  * Fixed handling of send()'s EAGAIN return value.

 -- Daniel Fiser <daniel.fiser@firma.seznam.cz>  Wed, 20 Jun 2012 11:22:46 +0200

libsphinxclient (1.3.1-3) unstable; urgency=low

  * Added EINPROGRESS error handling on reading from socket using recv(). 

 -- Jan Kirschner <jan.kirschner@firma.seznam.cz>  Thu, 15 Oct 2009 18:18:38 +0200

libsphinxclient (1.3.1-2) unstable; urgency=low

  * Added function addEnumFilter with IntArray for backward compatibility

 -- Jan Kirschner <jan.kirschner@firma.seznam.cz>  Fri, 03 Jul 2009 16:30:02 +0200

libsphinxclient (1.3.1-1) unstable; urgency=low

  * Corrected missing 64bit attribute parsing in response from sphinx

 -- Jan Kirschner <jan.kirschner@firma.seznam.cz>  Wed, 17 Jun 2009 13:50:40 +0200

libsphinxclient (1.3.0-1) unstable; urgency=low

  * Initial release - package creation reformed

 -- Jan Kirschner <jan.kirschner@firma.seznam.cz>  Wed, 17 Jun 2009 09:32:27 +0200



sphinxtest3.cc
/*
*
* C++ sphinx search client library
* Copyright (C) 2007  Seznam.cz, a.s.
*
* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU Lesser General Public
* License as published by the Free Software Foundation; either
* version 2.1 of the License, or (at your option) any later version.
*
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with this library; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*
* Seznam.cz, a.s.
* Radlicka 2, Praha 5, 15000, Czech Republic
* http://www.seznam.cz, mailto:sphinxclient@firma.seznam.cz
*
* $Id: sphinxtest.cc 15 2009-06-17 06:40:44Z honkir $
*
* DESCRIPTION
* A sphinxclient sample and testing program.
*
* AUTHORS
* Jan Kirschner <jan.kirschner@firma.seznam.cz>
*
* HISTORY
* 2006-12-14  (kirschner)
*             Created.
*
* Quick compile:
* g++  -g sphinxtest.cc -Iinclude/  -Lsrc/.libs/ -lsphinxclient -o sphinxtest
*/

#include <sphinxclient/sphinxclient.h>
#include <stdio.h>

//------------------------------------------------------------------------------


int main()
{
    Sphinx::ConnectionConfig_t config (
        "localhost", // hostname
        3312,        // port
        true,        // keepalive
        2000,        // connect timeout
        20000,       // read timeout
        2000         // write timeout
    );

    Sphinx::Client_t connection(config);
    std::vector<Sphinx::Response_t> results;
    Sphinx::SearchConfig_t settings(Sphinx::VER_COMMAND_SEARCH_0_9_9);

    printf("starting.....\n");

    // search settings
    settings.setPaging(0, 20);
    settings.setMatchMode(Sphinx::SPH_MATCH_ALL);
    settings.setMaxMatches(1000);
    settings.setSearchedIndexes("test1");
    settings.setQueryComment("comment");
    settings.setMaxQueryTime(10000);

    // preparation of enumeration filter
    Sphinx::Int64Array_t ids;
    ids.push_back(23);
    ids.push_back(10);

    // inicialise range and enum filter
    // outside of interval 50-70, inside of interval 11-60
    settings.addRangeFilter("att_uint", 50, 60, true);
    settings.addRangeFilter("att_uint", 11, 100, false);
    // skip document id in ids
    settings.addEnumFilter("att_uint", ids, true);

    // sorting - relevance ascending
    settings.setSorting(Sphinx::SPH_SORT_RELEVANCE);

    // groups settings
    settings.setGrouping(Sphinx::SPH_GROUPBY_ATTR, "att_group", "@id asc");

    // create multiquery
    Sphinx::MultiQueryOpt_t query(Sphinx::VER_COMMAND_SEARCH_0_9_9);
    query.addQuery("ped ahoj popisek", settings);
    query.addQuery("pes ahoj", settings);
    settings.setGrouping(Sphinx::SPH_GROUPBY_ATTR, "att_uint", "@id asc");
    query.addQuery("pes ahoj", settings);
    query.addQuery("ped ahoj popisek", settings);

    // group queries that are efficient for sphinx multi-query processing
    query.optimise();

    try{
        //multi query
        connection.query(query, results);
        //single query
        //connection.query("pes ahoj", settings, result);
        printf("query success.\n");
    }
    catch(Sphinx::Error_t e)
    {
        printf("query error:\n%s\n", e.errMsg.c_str());
        return 2;
    }

    printf("result count:      %d\n", results.size());
    if(results.size())
        printf("command version: 0x%X\n", results[0].commandVersion);

    for(int resId=0 ; resId<(int)results.size() ; resId++)
    {
        printf("------------------------ result %03d ---------------------\n", resId+1);
        Sphinx::Response_t &result = results[resId];
        printf("field count:       %d\n", result.field.size());
        printf("attribute count:   %d\n", result.attribute.size());
        printf("match count:       %d\n", result.entry.size());
        printf("word count:        %d\n", result.word.size());

        printf("\nFields:\n");
        for (int i=0;i<result.field.size();i++)
        {
            printf("    %s\n", result.field[i].c_str());
        }//for

        printf("\nAttributes:\n");
        for (Sphinx::AttributeTypes_t::iterator it=result.attribute.begin() ;
             it!=result.attribute.end();it++)
        {
            if (it->second & Sphinx::SPH_ATTR_MULTI) {
                printf("    name: %s, multi type: %d\n", it->first.c_str(), (it->second ^ Sphinx::SPH_ATTR_MULTI));
            }//if
            else
                printf("    name: %s, type: %d\n", it->first.c_str(), it->second);
        }//for

        printf("\nWords:\n");
        for (std::map<std::string, Sphinx::WordStatistics_t>::iterator it
             = result.word.begin() ; it!=result.word.end() ; it++)
        {
            printf("    Word %s: %d docs / %d hits\n", it->first.c_str(), it->second.docsHit, it->second.totalHits);
        }//for

        printf("entries: %d\n", result.entriesGot);
        printf("documents: %d\n", result.entriesFound);
        printf("duration: %d\n", result.timeConsumed);
        printf("64bit ID: %d\n", result.use64bitId);

        // print results with attributes
        printf("\nResponse:\n");

        int i = 0; 
        for (std::vector<Sphinx::ResponseEntry_t>::const_iterator it = result.entry.begin();
             it != result.entry.end(); it++) {

            i++;
            printf("%d) id: %ld ", i, it->documentId);
            for (std::map<std::string, Sphinx::Value_t>::const_iterator j =
                                it->attribute.begin();
                    j != it->attribute.end() ; j++)
            {
                switch(j->second.getValueType()) {
                case Sphinx::VALUETYPE_FLOAT:
                    printf("%s:%.2f | ", j->first.c_str(), (float)(j->second));
                    break;
                case Sphinx::VALUETYPE_VECTOR: {
                    char comma[2] = {'\0', '\0'};
                    std::vector<Sphinx::Value_t> vec = j->second;
                    printf("%s:(", j->first.c_str());
                    for (std::vector<Sphinx::Value_t>::iterator vi=vec.begin() ; vi!=vec.end() ; vi++) {
                        if(vi->getValueType() == Sphinx::VALUETYPE_FLOAT)
                            printf("%s%ff", comma, (float)(*vi));
                        else
                            printf("%s%dd", comma, (uint32_t)(*vi));
                        comma[0] = ',';
                    }//for
                    printf(") | ");
                    break; }
                case Sphinx::VALUETYPE_UINT64:
                    printf("%s:%lu | ", j->first.c_str(), (uint64_t)(j->second));
                    break;
                case Sphinx::VALUETYPE_UINT32:
                default:
                    printf("%s:%d | ", j->first.c_str(), (uint32_t)(j->second));
                    break;
                }//switch
            }
            printf("\n");
        }//for response entries
    }//for response

    printf("----------------------------- end ------------------------\n");

    return 0;
}//main

